### PR TITLE
Add support for ZADD options (NX, XX, CH, INCR)

### LINKED ISSUE #545: ZADD options are not supported
While for instance method `set` supports all SET command options, method `zadd` doesn't offer support for any ZADD options (NX, XX, CH, INCR).

### PR DESCRIPTION
**Pull request by jcmfernandes**
Closes #545

### CODE REVIEW

**Comment by jcmfernandes** [comment]
Here it is @yaauie. Happy to hear feedback. Cheers!

**Comment by jcmfernandes** [comment]
CI showed that my code is not Ruby 1.8.x compatible. I can make the `options` argument of `zadd` part of its variadic argument `*args`.

I would still prefer to consider `options` as parameter by itself and leave documentation accordingly but I'm more than happy to hear feedback on this one.

**Comment by jcmfernandes** [comment]
Anyone? :cry:

**Comment by gmalkas** [comment]
Any change to get this merged in? :)

**Comment by djanowski** [comment]
Sorry for the delay.

File: lib/redis.rb [hunk 1]
```diff
@@ -1410,20 +1410,53 @@ def zcard(key)
...
+    if args.last.is_a?(Hash)
+      options = args.pop
+
+      nx = options[:nx]
+      zadd_options << "NX" if nx
+
+      xx = options[:xx]
+      zadd_options << "XX" if xx
+
+      raise ArgumentError, "XX and NX options at the same time are not compatible" if nx && xx
```
Comment (djanowski) [inline lib/redis.rb]: We usually don't validate the semantics of commands and defer that to Redis itself. Can you please remove this before merging?

Comment (jcmfernandes) [inline lib/redis.rb]: Will do so.

File: lib/redis.rb [hunk 2]
```diff
@@ -1410,20 +1410,53 @@ def zcard(key)
...
-        client.call([:zadd, key] + args[0])
+        raise ArgumentError, "only one score-element pair can be specified with the :incr option" if incr
+        client.call([:zadd, key] + zadd_options + args[0])
       elsif args.size == 2
-        # Single pair: return boolean
-        client.call([:zadd, key, args[0], args[1]], &_boolify)
+        # Single pair: return float if INCR, boolean if !INCR
+        client.call([:zadd, key] + zadd_options + [args[0], args[1]], &(incr ? _floatify : _boolify))
       else
         raise ArgumentError, "wrong number of arguments"
```
Comment (jcmfernandes) [inline lib/redis.rb]: Should I remove this one too @djanowski ?

**Comment by jcmfernandes** [comment]
Ready for some more feedback.

**Comment by djanowski** [comment]
@jcmfernandes I know this is not your part of your PR, but can you help me get the build passing on CI?

**Comment by jcmfernandes** [comment]
It seems to me that the redis unstable branch is currently broken.... not much I can do about that @djanowski.

**Comment by jcmfernandes** [comment]
Still failing on redis unstable :crying_cat_face:.

```
  1) Failure:
test_hincrbyfloat(TestCommandsOnHashes) [/Users/jcf/Documents/talkdesk/redis-rb/test/lint/hashes.rb:150]:
<"1.23"> expected but was
<"1.22999999999999998">.

  2) Failure:
test_hincrbyfloat(TestDistributedCommandsOnHashes) [/Users/jcf/Documents/talkdesk/redis-rb/test/lint/hashes.rb:150]:
<"1.23"> expected but was
<"1.22999999999999998">.

568 tests, 5155 assertions, 2 failures, 0 errors, 0 skips
```

**Comment by badboy** [comment]
The float formatting problem was fixed upstream.

**Comment by jcmfernandes** [comment]
Thanks @badboy. @djanowski all green now.

**Comment by djanowski** [comment]
Thank you.

**Comment by djanowski** [comment]
This is available in 3.2.2.

**Pull request merged**